src/Npgsql/TypeHandlers/ArrayHandler.cs

```diff
@@ -91,21 +91,59 @@ public ArrayHandler([CanBeNull] NpgsqlTypeHandler elementHandler, int lowerBound
+                var elementType = t.GetElementType();
+                // We need to treat this as an actual array type, these need special treatment because of
+                // typing/generics reasons (there is no way to express "array of X" with generics
+                if (elementType == GetElementFieldType())
+                    return (TArray)await ReadAsObject(buf, len, async, fieldDescription);
+                if (elementType == GetElementPsvType())
+                    return (TArray)await ReadPsvAsObject(buf, len, async, fieldDescription);
+            }
+
+            if (t == typeof(List<TElement>))
+                return (TArray)await ReadAsList<TElement>(buf, async);
+
+            if (t.IsGenericType
+                && t.GetGenericTypeDefinition() == typeof(List<>)
+                && t.GenericTypeArguments[0] == GetElementPsvType())
+            {
+                return (TArray)await ReadPsvAsList(buf, async);
+            }
+                throw new InvalidCastException($"Can't cast database type {PgDisplayName} to {typeof(TArray).Name}");
+        }
+
+        protected async ValueTask<object> ReadAsList<TElement2>(NpgsqlReadBuffer buf, bool async)
+        {
+            await buf.Ensure(12, async);
+            var dimensions = buf.ReadInt32();
+
+            if (dimensions == 0)
+                return new List<TElement2>();
+            if (dimensions > 1)
+                throw new NotSupportedException($"Can't read multidimensional array as List<{typeof(TElement2).Name}>");
+
+            buf.ReadInt32();        // Has nulls. Not populated by PG?
+            var elementOID = buf.ReadUInt32();
+            // The following should hold but fails in test CopyTests.ReadBitString
+            //Debug.Assert(elementOID == ElementHandler.BackendType.OID);
+
+            await buf.Ensure(8, async);
+            var length = buf.ReadInt32();
+            buf.ReadInt32(); // We don't care about the lower bounds
 
-            // We need to treat this as an actual array type, these need special treatment because of
-            // typing/generics reasons (there is no way to express "array of X" with generics
-            var elementType = t.GetElementType();
-            if (elementType == GetElementFieldType())
-                return (TArray)await ReadAsObject(buf, len, async, fieldDescription);
-            if (elementType == GetElementPsvType())
-                return (TArray)await ReadPsvAsObject(buf, len, async, fieldDescription);
-            throw new InvalidCastException($"Can't cast database type {PgDisplayName} to {typeof(TArray).Name}");
+            var list = new List<TElement2>(length);
+            for (var i = 0; i < length; i++)
+                list.Add(await ElementHandler.ReadWithLength<TElement2>(buf, async));
+            return list;
         }
 
+        protected internal virtual ValueTask<object> ReadPsvAsList(NpgsqlReadBuffer buf, bool async)
+            => throw new NotSupportedException();
```

Comment:
Remove. See the comment above.

Reply:
I don't understand what you mean here. IMO removing this method would mean removing the functionality of reading arrays as List of provider specific values.